docs/examples/basic-options.md
---
jupytext:
  text_representation:
    extension: .md
    format_name: myst
    format_version: 0.13
    jupytext_version: 1.16.4
kernelspec:
  display_name: Python 3
  language: python
  name: python3
---

# Basic options

This demo explains the basic and required options that you need to set to use psignifit.

```{code-cell} ipython3
import numpy as np

import psignifit as ps

# to have some data we use the data from the first demo.
data = np.array([[0.0010, 45.0000, 90.0000], [0.0015, 50.0000, 90.0000],
                 [0.0020, 44.0000, 90.0000], [0.0025, 44.0000, 90.0000],
                 [0.0030, 52.0000, 90.0000], [0.0035, 53.0000, 90.0000],
                 [0.0040, 62.0000, 90.0000], [0.0045, 64.0000, 90.0000],
                 [0.0050, 76.0000, 90.0000], [0.0060, 79.0000, 90.0000],
                 [0.0070, 88.0000, 90.0000], [0.0080, 90.0000, 90.0000],
                 [0.0100, 90.0000, 90.0000]])

res = ps.psignifit(data, sigmoid='norm', experiment_type='2AFC')
```

## Experiment type

Depending on the experimental condition, the experimenter has different a priori knowledge about the asymptotes. In *psignifit* this is handled by an option called `experiment_type`. There are three different types available:

### nAFC Experiments
The first is meant for n-alternative-forced-choice experiments, i. e. for experiments for which n alternatives are given, an answer is enforced and exactly one alternative is right. You can choose it by setting the option `experiment_type` to the string "nAFC", where you replace *n* with the number of options. For example, if you use 2, 3 or 4 alternatives you may pass:

```
experiment_type = '2AFC'
experiment_type = '3AFC'
experiment_type = '4AFC'
```

This mode fixes the lower asymptote gamma to $1/n$ and leaves the upper asymptote free to vary.

### Yes/No Experiments ###
Intended for simple detection experiments asking subjects whether they perceived a single presented stimulus or not, or any other experiment which has two possible answers of which one is reached for "high" stimulus levels and the other for "low" stimulus levels. You choose it with:

```
experiment_type = 'yes/no'
```

This sets both asymptotes free to vary and applies a prior to them favoring small values, e.g. asymptotes near 0 and 1 respectively.

### Equal Asymptote Experiments ###
This setting is essentially a special case of Yes/No experiments. Here the asymptotes are "yoked", i. e. they are assumed to be equally far from 0 or 1. This corresponds to the assumption that stimulus independent errors are equally likely for clear "Yes" answers as for clear "No" answers. It is chosen with:

```
experiment_type = 'equal asymptote'
```

Note that this will make fitting the psychometric function considerably faster than `yes/no` because *psignifit* has to only fit four rather than five parameters.


## Sigmoid


The default value of `sigmoid` is the cumulative Gaussian. 
```
sigmoid = 'norm'
```

Another standard alternative is the logistic function.

```
sigmoid = 'logistic'
```

The Gumbel and reversed Gumbel functions for asymmetric psychometric functions are also included. The Gumbel has a longer lower tail the reversed Gumbel a longer upper tail.  

```
sigmoid = 'gumbel'
sigmoid = 'rgumbel'
```

You can find [all sigmoids supported by psignifit here](plot_all_sigmoids).

If the standard sigmoids do not match your requirements, you may provide your own sigmoid, which has to be a subclass of `psignifit.sigmoids.Sigmoid`. 
Consult the [API reference](../reference/api) and the source code to learn how to set your own sigmoid.



docs/examples/advanced-options.md
---
jupytext:
  text_representation:
    extension: .md
    format_name: myst
    format_version: 0.13
    jupytext_version: 1.16.4
kernelspec:
  display_name: Python 3
  language: python
  name: python3
---

# Advanced options

This demo explains the different advanced options that you can set as a user in *psignifit*.

```{code-cell} ipython3
import numpy as np

import psignifit as ps

# to have some data we use the data from the first demo.
data = np.array([[0.0010, 45.0000, 90.0000], [0.0015, 50.0000, 90.0000],
                 [0.0020, 44.0000, 90.0000], [0.0025, 44.0000, 90.0000],
                 [0.0030, 52.0000, 90.0000], [0.0035, 53.0000, 90.0000],
                 [0.0040, 62.0000, 90.0000], [0.0045, 64.0000, 90.0000],
                 [0.0050, 76.0000, 90.0000], [0.0060, 79.0000, 90.0000],
                 [0.0070, 88.0000, 90.0000], [0.0080, 90.0000, 90.0000],
                 [0.0100, 90.0000, 90.0000]])

res = ps.psignifit(data, sigmoid='norm', experiment_type='2AFC')
```


## Estimation Type
As described in detail in the section [Parameters estimates: MAP and mean](map_vs_mean), this option sets how you want to estimate your fit from the posterior.

- 'mean': using the posterior mean. In a Bayesian sense this is a more suitable estimate of the expected value of the posterior. However, in our paper we show that for psychometric function fitting and the small datasets used in experimental psychology and the neurosciences, and for the our priors the MAP is preferable as default!
- 'MAP': maximum a posteriori computed from the posterior. Default.

```{code-cell} ipython3
# This gives the default estimate, in this case the mean estimate since we set it in the options.
print(f"parameter estimate (default=MAP): {res.parameter_estimate}")
print(f"parameter estimate (MAP): {res.parameter_estimate_MAP}")
print(f"parameter estimate (mean){res.parameter_estimate_mean}")
```

you can also get it using this method:
```{code-cell} ipython3
print(res.get_parameter_estimate(estimate_type="MAP"))
```

## Optimization Steps
This option sets the number of grid points on each dimension in the final
fitting (`grid_steps`) and in the moving of bounds (`steps_moving_bounds`).
The number of steps is passed as a dictionary with keys `threshold`, `width`, `lambda` (upper asymptote),
`gamma` (lower asymptote), and `eta` (overdispersion parameter).

You may change this if you need more accurate estimates on the sparsely
sampled parameters. For example, to get an even more exact estimate on the
lapse rate/upper asymptote, set lambda to 50.
In this case the lapse rate is sampled at 50 points giving you a much more exact and smooth curve for comparisons.

```{code-cell} ipython3
res = ps.psignifit(data, grid_steps={'lambda': 50})
```

## Confidence intervals

### Confidence level
You can also personalize the confidence level for the computed confidence intervals.
This may be set to any number between 0 and 1 excluding.

For example to get 95% **and** 99% confidence intervals try

```{code-cell} ipython3
res = ps.psignifit(data, confP=(.95, .99))
```

### Computing method

The option `CI_method` sets how the confidence intervals are computed. The possible options are:

- `project`: project the confidence region from the multidimensional posterior grid on each axis.

- `percentiles`: find $\alpha/2$ and $(1-\alpha/2)$ percentiles (with $\alpha$ = 1 - confP). Default.

```{code-cell} ipython3
res = ps.psignifit(data, CI_method='percentiles')
# or
res = ps.psignifit(data, CI_method='project')
```

## Threshold definition (`thresh_PC`)

This option sets the proportion correct to correspond to the threshold on the *unscaled* sigmoid. Possible values 
are in the range from 0 to 1, default is 0.5. The default corresponds to 75\% in a 2AFC task (midway between the 
guess rate of 50 % and ceiling performance 100%).

To set it to a different value, for example to 90 %, you'll do

```{code-cell} ipython3
res = ps.psignifit(data, thresh_PC=0.9)
```

Note that this corresponds to a 95 \% in a 2AFC task.

Note also that you will get a warning, because the default prior assumes that the experimental
stimulus range covers the range where the threshold likely falls. If this doesn't match your
setup, you'll need a custom prior. See the [priors demo](priors) to learn how to do it.



## Width definition (`width_alpha`)

The definition of the `width` parameter of a psychometric function can be changed with the option `width_alpha`.

`width = `$\Psi^{-1}(1-\alpha) - \Psi^{-1}(\alpha)$ where $\Psi^{-1}$ is the inverse of the sigmoid function.

`width_alpha` must be between 0 and .5 excluding. Default is `width=0.05`

For example, this would enable the usage of the interval from .1 to .9 as the width

```{code-cell} ipython3
res = ps.psignifit(data, width_alpha=0.05)
```

## Priors

You can set the priors manually passing them as a dictionary with the parameter name as key and the custom prior function
as the value.

For details on how do change these refer to the [priors demo](priors).


The strength of the prior in favor of a binomial observer can be set with
the `beta_prior` option. Larger values correspond to a stronger prior.
The default value is 10; we choose this value after
a rather large number of simulations. Refer to
[Schütt & Wichmann (2016)](http://www.sciencedirect.com/science/article/pii/S0042698916000390) 
to learn more about this.

```{code-cell} ipython3
res = ps.psignifit(data, beta_prior=15)
```


## Bounds

```{warning}
By changing the default bounds you might artificially bias the estimation procedure. 
The defaults were choosen after extensive simulations; these are reported in 
[Schütt & Wichmann (2016)](http://www.sciencedirect.com/science/article/pii/S0042698916000390). 
Change the defaults if you know what you are doing!
```

You may provide your own bounds for the parameters.
This should be a dictionary of the parameter name and tuples of
start and end of the range.

For example, the following setting will set the bounds for the threshold between 0 and 0.002,
for the lapse rate (lambda) between 0 and 0.1, and a fixed guess rate (gamma) to 0.5

```{code-cell} ipython3
custom_bounds = {
                 'threshold': (0.0, 0.015),
                 'lambda': (0.0, 0.1),
                 'gamma': (.5, .5),
                }
res = ps.psignifit(data, bounds=custom_bounds)
```


Parts of the grid which produce marginal values below a small value are rounded to zero 
and excluded in further calculations. This small value is set in parameter `max_bound_values`.
It should be a very small value and at least smaller than `1/(max(grid_steps))`.

This for example would exclude fewer values and more conservative
movement of the bounds:

```{code-cell} ipython3
res = ps.psignifit(data, max_bound_value=np.exp(-20))
```


psignifit/configuration.py
"""This module defines the basic configuration object for psignifit.
"""
import re
import dataclasses
from typing import Any, Dict, Tuple, Optional, Union

from . import sigmoids
from .utils import PsignifitException
from .typing import ExperimentType, Prior


_PARAMETERS = {'threshold', 'width', 'lambda', 'gamma', 'eta'}


@dataclasses.dataclass
class Configuration:
    """The basic configuration object for psignifit.

    This class contains a set of valid options and the corresponding sanity
    checks.
    These checks raise `PsignifitException` if an invalid option is specified or
    if a valid option is specified with a value outside of the allowed range.

    The sanity checks are only run during initialization.
    Changing attributes is highly discouraged and should be followed
    by rerunning sanity checks with `config.check_attributes()`.

    Note: attributes and methods starting with `_` are considered private and
    used internally. Do not change them unlsess you know what you are doing

    Note for the developer: if you want to add a new valid option `foobar`,
    expand the `Conf.valid_opts` tuple (in alphabetical order) and add any
    check in a newly defined method `def check_foobar(self, value)`, which
    raises `PsignifitException` if `value` is outside of the accepted range
    for `foobar`.
    """
    beta_prior: int = 10
    CI_method: str = 'project'
    confP: Tuple[float, float, float] = (.95, .9, .68)
    estimate_type: str = 'MAP'
    experiment_type: str = ExperimentType.YES_NO.value
    experiment_choices: Optional[int] = None
    fast_optim: bool = False
    fixed_parameters: Optional[Dict[str, float]] = None
    grid_set_type: str = 'cumDist'
    instant_plot: bool = False
    max_bound_value: float = 1e-05
    move_bounds: bool = True
    pool_max_blocks: int = 25
    priors: Optional[Dict[str, Prior]] = dataclasses.field(default=None, hash=False)
    sigmoid: Union[str, sigmoids.Sigmoid] = 'norm'
    stimulus_range: Optional[Tuple[float, float]] = None
    thresh_PC: float = 0.5
    verbose: bool = True
    width_alpha: float = 0.05
    width_min: Optional[float] = None

    # attributes, if not specified, will be initialize based on others
    bounds: Optional[Dict[str, Tuple[float, float]]] = None
    grid_steps: Dict[str, int] = dataclasses.field(default_factory=dict)
    steps_moving_bounds: Dict[str, int] = dataclasses.field(default_factory=dict)


    def __post_init__(self):
        self.check_attributes()

    @classmethod
    def from_dict(cls, config_dict: Dict[str, Any]):
        config_dict = config_dict.copy()
        return cls(confP=tuple(config_dict.pop('confP')),
                   **config_dict)

    def as_dict(self) -> Dict[str, Any]:
        return dataclasses.asdict(self)

    def check_attributes(self):
        """ Run sanity checks.

        For each attribute named NAME, this function
        searches for a method check_NAME and executes
        it if found.

        Example for a sanity check of a `foobar` attribute:
        .. code-block:: python

            def check_foobar(self, value):
                if value > 10:
                    raise PsignifitException(f'Foobar must be < 10: {value} given!')
         """
        for attribute in dataclasses.fields(self):
            sanity_check_name = 'check_' + attribute.name
            if hasattr(self, sanity_check_name):
                sanity_check_method = getattr(self, sanity_check_name)
                attribute_value = getattr(self, attribute.name)
                sanity_check_method(attribute_value)

    def check_bounds(self, value):
        if value is not None:
            # bounds is a dict in the form {'parameter_name': (left, right)}
            if type(value) != dict:
                raise PsignifitException(
                    f'Option bounds must be a dictionary ({type(value).__name__} given)!'
                )
            # now check that the keys in the dictionary are valid
            vkeys = _PARAMETERS
            if vkeys < set(value.keys()):
                raise PsignifitException(
                    f'Option bounds keys must be in {vkeys}. Given {list(value.keys())}!'
                )
            # now check that the values are sequences of length 2
            for v in value.values():
                try:
                    correct_length = (len(v) == 2)
                except Exception:
                    correct_length = False
                if not correct_length:
                    raise PsignifitException(
                        f'Bounds must be a sequence of 2 items: {v} given!')

    def check_fixed_parameters(self, value):
        if value is not None:
            # fixed parameters is a dict in the form {'parameter_name': value}
            if type(value) != dict:
                raise PsignifitException(
                    f'Option fixed_parameters must be a dictionary ({type(value).__name__} given)!'
                )
            # now check that the keys in the dictionary are valid
            vkeys = _PARAMETERS
            if vkeys < set(value.keys()):
                raise PsignifitException(
                    f'Option fixed_paramters keys must be in {vkeys}. Given {list(value.keys())}!'
                )

    def check_experiment_type(self, value):
        valid_values = [type.value for type in ExperimentType]
        is_valid = value in valid_values
        is_nafc = re.match('[0-9]+AFC', value)
        if not (is_valid or is_nafc):
            raise PsignifitException(
                f'Invalid experiment type: "{value}"\nValid types: {valid_values},' +
                ', or "2AFC", "3AFC", etc...')
        if is_nafc:
            self.experiment_choices = int(value[:-3])
            self.experiment_type = ExperimentType.N_AFC.value
            value = ExperimentType.N_AFC.value
        if value == ExperimentType.N_AFC.value and self.experiment_choices is None:
            raise PsignifitException("For nAFC experiments, expects 'experiment_choices' to be a number, got None.\n"
                                     "Can be specified in the experiment type, e.g. 2AFC, 3AFC, … .")

        default_grid_steps = {
            'threshold': 40,
            'width': 40,
            'lambda': 20,
            # 'gamma' will be set below
            'eta': 20,
        }
        if value == ExperimentType.YES_NO.value:
            default_grid_steps['gamma'] = 20
            self.steps_moving_bounds = {
                'threshold': 25,
                'width': 30,
                'lambda': 10,
                'gamma': 10,
                'eta': 15,
                **self.steps_moving_bounds
            }
        else:
            default_grid_steps['gamma'] = 1
            self.steps_moving_bounds = {
                'threshold': 30,
                'width': 40,
                'lambda': 10,
                'gamma': 1,
                'eta': 20,
                **self.steps_moving_bounds
            }
        self.grid_steps = {**default_grid_steps,
                           **self.grid_steps}

    def check_sigmoid(self, value):
        try:
            self.make_sigmoid()
        except KeyError:
            raise PsignifitException('Invalid sigmoid name "{value}", use one of {sigmoids.ALL_SIGMOID_NAMES}')

    def check_stimulus_range(self, value):
        if value:
            try:
                len_ = len(value)
                wrong_type = False
            except TypeError:
                wrong_type = True
            if wrong_type or len_ != 2:
                raise PsignifitException(
                    "Option stimulus range must be a sequence of two items!")

    def check_width_alpha(self, value):
        try:
            # check that it is a number:
            diff = 1 - value
            wrong_type = False
        except Exception:
            wrong_type = True
        if wrong_type or not (0 < diff < 1):
            raise PsignifitException(
                f"Option width_alpha must be between 0 and 1 ({value} given)!")

    def check_width_min(self, value):
        if value:
            try:
                _ = value + 1
            except Exception:
                raise PsignifitException("Option width_min must be a number")

    def make_sigmoid(self) -> sigmoids.Sigmoid:
        """ Construct sigmoid according to this configuration.

        Returns:
             Sigmoid object with percentage correct and alpha according to config.
        """
        if isinstance(self.sigmoid, sigmoids.Sigmoid):
            self.sigmoid.PC = self.thresh_PC
            self.sigmoid.alpha = self.width_alpha
            return self.sigmoid
        else:
            return sigmoids.sigmoid_by_name(self.sigmoid, PC=self.thresh_PC, alpha=self.width_alpha)
